File: chap2/sec2.3/Parareal.m
function U=Parareal(F,G,T,u0,N,K);
% PARAREAL implementation of the parareal algorithm
%   U=Parareal(F,G,T,u0,N,K); applies the parareal algorithm with fine
%   solver F(t0,t1,ut0) and coarse solver G(t0,t1,ut0) on [0,T] with
%   initial condition u0 at t=0 using N equidistant coarse time points
%   doing K iterations. The output U{k} contains the parareal
%   approximations at the coarse time points for each iteration k.

dT=T/N; TT=0:dT:T;                               % coarse time mesh
U{1}(1,:)=u0;                                    
for n=1:N                                        % initial guess with G 
  Go(n+1,:)=G(TT(n),TT(n+1),U{1}(n,:));         
  U{1}(n+1,:)=Go(n+1,:);                         % keep Go for parareal 
end;
for k=1:K                                        % parareal iteration  
  for n=1:N                                  
    Fn(n+1,:)=F(TT(n),TT(n+1),U{k}(n,:));        % parallel with F
  end;
  U{k+1}(1,:)=u0;
  for n=1:N                                  
    Gn(n+1,:)=G(TT(n),TT(n+1),U{k+1}(n,:));      % sequential with G
    U{k+1}(n+1,:)=Fn(n+1,:)+Gn(n+1,:)-Go(n+1,:); % parareal update 
  end;
  Go=Gn;                                         % keep for next iteration
end;

File: chap2/sec2.3/Parareal.py
import numpy as np

def parareal(F, G, T, u0, N, K):
    """
    Implements the Parareal algorithm.

    Parameters
    ----------
    F : callable
        Fine solver, F(t0, t1, u0), advancing from t0 to t1 with initial condition u0.
    G : callable
        Coarse solver, G(t0, t1, u0), advancing from t0 to t1 with initial condition u0.
    T : float
        Final time.
    u0 : array-like
        Initial condition at t = 0.
    N : int
        Number of equidistant coarse time points.
    K : int
        Number of Parareal iterations.

    Returns
    -------
    U : list of numpy.ndarray
        List where U[k] contains the Parareal approximations at the coarse time points for each iteration k.
    """
    u0 = np.asarray(u0)
    TT = np.linspace(0, T, N+1)         # Coarse time mesh
    U = np.zeros((K+1, N+1, u0.size), dtype=u0.dtype)

    U[:, 0] = u0        # Initial condition (for all iterations)

    for n in range(N):  # Initial guess
        U[0, n+1] = G(TT[n], TT[n + 1], U[0, n])

    for k in range(K):  # Parareal iterations
        for n in range(N):
            U[k+1, n+1] = F(TT[n], TT[n+1], U[k, n]) \
                + G(TT[n], TT[n+1], U[k+1, n]) - G(TT[n], TT[n+1], U[k, n])

    return np.squeeze(U)
